File: reference/rocketride-server/nodes/src/nodes/text_output/endpoint.py
# =============================================================================
# MIT License
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# =============================================================================

import errno
import re
from threading import Lock
from typing import Any
from zlib import crc32

import engLib
from engLib import Filters


class Endpoint:
    """General connection to integration target 'Text Output'."""

    def beginEndpoint(self):
        """Call from engLib, Endpoint initialization."""
        # Connect and open key-value storage only when action is running
        if self.task_type == 'action':
            self.connect()

    def endEndpoint(self):
        """Call from engLib, Endpoint cleanup."""
        pass

    def validateConfig(self, syntaxOnly: bool):
        """Call from engLib, Endpoint configuration validation.

        Args:
            syntaxOnly (bool): if true, then do not connect, but only validate the syntax of parameters.
        """
        try:
            # RFC-1123
            IP_REGEX = r'^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$'
            HOSTNAME_REGEX = r'^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$'
            SERVER_REGEX = f'{IP_REGEX}|{HOSTNAME_REGEX}'
            Endpoint.validate_param('Server Name', self.server, pattern=SERVER_REGEX)

            Endpoint.validate_param(
                'User Name', self.username, is_required=bool(self.password), validator=Endpoint.validate_smb_username
            )
            Endpoint.validate_param('Password', self.password, is_required=bool(self.username), max_length=127)
            Endpoint.validate_param(
                'Store Path', self.store_path, min_lenght=3, max_length=256, validator=Endpoint.validate_smb_path
            )
            Endpoint.validate_param(
                'Anonymization Character', self.anonymize_char, is_required=self.anonymize, min_lenght=1, max_length=1
            )

            if not syntaxOnly:
                # Do not try to connect as validation is run by Platform
                # where SMB share may not be available
                # self.connect()
                pass

        except ValueError as e:
            engLib.error(e)

    def getConfigSubKey(self):
        """Call from engLib, target service uniqueness key."""
        if self.endpoint.serviceMode == 1:  # source mode
            raise Exception('Source mode not supported')
        else:  # target mode
            return f'{self.server}/{self.store_path}'.lower()

    def getPipeFilters(self):
        """Call from engLib, list of pipe filters."""
        if self.anonymize:
            self.endpoint.taskConfig['wantsPolicies'] = False
            self.endpoint.taskConfig['wantsContext'] = False
            self.endpoint.taskConfig['wantsText'] = True
            return [Filters.CLASSIFY, 'anonymize_text']

        else:
            return []

    def connect(self):
        """Authenticate and test connection to the target share.

        according to the configuration parameters.

        Raise:
            An exception if connection failed.
        """
        import smbclient
        from smbprotocol.exceptions import SMBOSError

        # Setup user name and password
        if self.username and self.password:
            smbclient.ClientConfig(username=self.username, password=self.password)

        # Build paths to check
        share = next(iter(re.split(r'[\\/]', self.store_path)), None)
        server_path = f'//{self.server}/{share}'
        server_store_path = f'//{self.server}/{self.store_path}'

        # Server path should be available
        smbclient.stat(server_path)

        try:
            # Check full store path for existence
            smbclient.stat(server_store_path)
        except SMBOSError as e:
            # Re-raise if it is not 'File not found'
            if e.errno != errno.ENOENT:
                raise

        engLib.debug('Connected', server_path)

    @staticmethod
    def validate_param(
        param_name: str,
        param_value: str,
        is_required: bool = True,
        min_lenght: int = None,
        max_length: int = None,
        pattern: str = None,
        validator: Any = None,
    ) -> bool:
        """Validate parameter to match the requirements."""
        if not param_value:
            if is_required:
                raise ValueError(f'{param_name} not specified')
            else:
                return

        if min_lenght is not None and len(param_value) < min_lenght:
            raise ValueError(f'{param_name} too short')

        if max_length is not None and len(param_value) > max_length:
            raise ValueError(f'{param_name} too long')

        if pattern and not re.match(pattern, param_value):
            raise ValueError(f'{param_name} not valid')

        if validator:
            try:
                validator(param_value)
            except ValueError as e:
                raise ValueError(f'{param_name} {e}')

    @staticmethod
    def validate_smb_username(username: str):
        m = re.match(r'^(.+)\\(.+)$', username)
        if not m:
            raise ValueError('should be specified using domain-like format')
        # domain = m.group(1)
        # username = m.group(2)

    @staticmethod
    def validate_smb_path(smb_path: str):
        """Validate sub-path to be not rooted, not containing dot folders and special characters."""
        path_items = re.split(r'[\\/]', smb_path)
        for i, dir_name in enumerate(path_items):
            if i == 0:
                SHARE_REGEX = r"^[a-zA-Z0-9\s!@#$%&'_\-\.~\(\){}]{1,80}$"
                if not re.match(SHARE_REGEX, dir_name):
                    raise ValueError('Share Name not valid')
            if not dir_name and i < len(path_items) - 1:
                raise ValueError('should not be rooted' if i == 0 else 'should not contain empty folder')
            if dir_name == '.' or dir_name == '..':
                raise ValueError('should not contain dot-folders')
            if any(dir_name.find(c) != -1 for c in '<>:"|?*'):
                raise ValueError('contains invalid characters')

    def _settings_changed(self) -> bool:
        """Check classification settings and reset saved state if changed.

        Returns:
            * True if the settings have been changed
              and all objects are to be transformed,
              including unchanged objects from a previous transformation.
            * False if the settings have been changed and
              only new and changed objects need to be transformed.
        """
        if self.settings_changed_value is not None:
            return self.settings_changed_value

        with self.settings_changed_lock:
            if self.settings_changed_value is not None:
                return self.settings_changed_value

            # Current settings key
            settings_key = None
            if self.anonymize:
                # Calculate classify hash
                classify_hash = crc32(str(self.endpoint.taskConfig.get('classifyPolicies')).encode())
                settings_key = f'{classify_hash:x};{self.anonymize_char};{self.anonymize_all}'
            else:
                settings_key = ''

            # Last settings key
            prev_settings_key = self.endpoint.keystore.getValue(':settings-key:')

            # If settings changed...
            if settings_key != prev_settings_key:
                # Update settings key
                self.endpoint.keystore.setValue(':settings-key:', settings_key)

                self.settings_changed_value = True

            else:
                self.settings_changed_value = False

            return self.settings_changed_value

    #
    # Configuration parameters
    #
    task_type = property(lambda self: self.endpoint.jobConfig.get('type'))
    server = property(lambda self: self.endpoint.parameters.get('server'))
    username = property(lambda self: self.endpoint.parameters.get('username'))
    password = property(lambda self: self.endpoint.parameters.get('password'))
    anonymize = property(lambda self: self.endpoint.parameters.get('anonymize', False))
    anonymize_char = property(lambda self: self.endpoint.parameters.get('anonymizeChar', '\u2588'))
    anonymize_all = property(lambda self: self.endpoint.parameters.get('anonymizeAll', False))
    store_path = property(lambda self: self.endpoint.parameters.get('storePath'))

    settings_changed: bool = property(lambda self: self._settings_changed())
    """
    Whether or not the transform related settings have been changed
    * True to transform all objects
    * False to transform only new and changed objects
    """
    settings_changed_value: bool = None
    settings_changed_lock = Lock()

File: reference/rocketride-server/packages/server/engine-lib/engLib/store/core/endpoint/endpoint.cpp
// =============================================================================
// MIT License
// Copyright (c) 2026 Aparavi Software AG
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================

#include <engLib/eng.h>

namespace engine::store {
//-------------------------------------------------------------------------
/// @details
///		Destructor - ends the endpoint
//-------------------------------------------------------------------------
IServiceEndpoint::~IServiceEndpoint() {
    LOGPIPE();

    // Remove it if we are registered
    Debugger::deregisterDebugger(taskId);

    // Stop
    endEndpoint();
};

//-------------------------------------------------------------------------
/// @details
///		Sets up the endpoint
///	@param[in]	openMode
///		The open mode. Will be either READ or WRITE
///	@returns
///		Error
//-------------------------------------------------------------------------
Error IServiceEndpoint::beginEndpoint(OPEN_MODE _openMode) noexcept {
    Error ccode;

    LOGPIPE();

    // Should not be opened
    if (config.openMode != OPEN_MODE::NONE)
        return APERR(Ec::InvalidCommand, "Endpoint already open");

    // Set/reset open mode
    util::Guard pipes{[&] { config.openMode = _openMode; },
                      [&] {
                          if (ccode) config.openMode = OPEN_MODE::NONE;
                      }};

    // Get the pipe stack to create
    if (ccode = buildPipeStack()) return ccode;

    // If we are actually starting the endpoint, we need to build
    // the global pipe and the connectoions
    if (!this->m_stackOnly) {
        // Build the global pipe
        if (ccode = buildGlobalPipe()) return ccode;

        // We have now finalized all of our stack components, so we
        // can build the connection list
        if (ccode = buildConnections()) return ccode;

        // Build the breakpoint list
        if (ccode = buildBreakpoints()) return ccode;
    }

    // And done
    return {};
}

//-------------------------------------------------------------------------
/// @details
///		Sets up the endpoint
///	@param[in]	openMode
///		The open mode. Will be either READ or WRITE
///	@returns
///		Error
//-------------------------------------------------------------------------
Error IServiceEndpoint::signal(const Text &signal,
                               json::Value &param) noexcept {
    return {};
}

//-------------------------------------------------------------------------
/// @details
///		Destructs or ends the endpoint access
///	@returns
///		Error
//-------------------------------------------------------------------------
Error IServiceEndpoint::endEndpoint() noexcept {
    Error ccode;

    LOGPIPE();

    // Clear our selection info
    deinitSelections();

    // Clear the target (if any)
    target = {};

    // Destroy all the instance pipes and their filters we created,
    // even if we are not open
    while (!m_instanceStacks.empty()) {
        // Get the last stack
        auto instanceStack = _mv(m_instanceStacks.back());
        m_instanceStacks.pop_back();

        // For each filter in this instance stack
        while (!instanceStack.empty()) {
            // Get the last filter
            auto filter = _mv(instanceStack.back());
            instanceStack.pop_back();

            // Tell the filter to end
            ccode = filter->endFilterInstance() || ccode;
        }
    }

    // Destroy all the globals, even if we are not open
    while (!m_globalStack.empty()) {
        // Get the last filter
        auto filter = _mv(m_globalStack.back());
        m_globalStack.pop_back();

        // Close it if we are not in config mode
        if (config.openMode != OPEN_MODE::CONFIG &&
            config.openMode != OPEN_MODE::PIPELINE_CONFIG) {
            ccode = filter->endFilterGlobal() || ccode;
        }
    }

    // End it
    config.openMode = OPEN_MODE::NONE;
    return ccode;
}

//-------------------------------------------------------------------------
/// @details
///		Static function to easily get the major type of the endpoint
///	@param[in]	serviceSection
///		The service section
///	@returns
///		Error
//-------------------------------------------------------------------------
ErrorOr<Text> IServiceEndpoint::getLogicalType(
    const json::Value &serviceConfig) noexcept {
    // Look up the type
    Text type = serviceConfig.lookup<iText>("type").lowerCase();

    // Split off the :// if it is there
    const auto parsed = type.split(':');

    // Prevent a crash
    if (parsed.size() < 1)
        return APERR(Ec::InvalidCommand, "Endpoint type not specified");

    // Return the type without the ://
    return parsed[0];
}

//-------------------------------------------------------------------------
/// @details
///		Gets the physical node type. For example, given ms-ondrive,
///		this returns python, or given filesys, it will return filesys
///	@param[in]	serviceConfig
///		The service section
//-------------------------------------------------------------------------
ErrorOr<Text> IServiceEndpoint::getPhysicalType(
    const json::Value &serviceConfig) noexcept {
    // Get the type specified
    auto type = IServiceEndpoint::getLogicalType(serviceConfig);
    if (!type) return type.ccode();

    // Get our service definition if we have one, if we don't just
    // return the logical type
    auto service = IServices::getServiceDefinition(*type);
    if (!service) return service.ccode();

    // Return the physical type
    return (*service)->physicalType;
}

//-------------------------------------------------------------------------
/// @details
///		Static function to easily create an endpoint
/// @param[in]	endpointMode
///		The mode of the endpoint
///	@param[in]	param
///		The parameters for the endpoint
//-------------------------------------------------------------------------
ErrorOr<ServiceEndpoint> IServiceEndpoint::getEndpoint(
    ENDPOINT_MODE endpointMode, ENDPOINT_PARAM &param) noexcept {
    // If there is no service config
    auto defaultServiceConfig = json::Value();

    // Based on the mode
    switch (endpointMode) {
        case ENDPOINT_MODE::SOURCE: {
            // We need a source endpoint - generate it
            defaultServiceConfig = R"(
                        {
                            "key": "null://*",
                            "name": "Null endpoint",
                            "type": "null",
                            "mode": "source",
                            "parameters": {}
                        }
                    )"_json;
            break;
        }

        case ENDPOINT_MODE::TARGET: {
            // We need a target endpoint - generate it
            defaultServiceConfig = R"(
                        {
                            "key": "null://*",
                            "name": "Null endpoint",
                            "type": "null",
                            "mode": "target",
                            "parameters": {}
                        }
                    )"_json;
        }
    }

    // Check the job config
    if (!param.jobConfig.has_value() || param.jobConfig.value().get().isNull())
        return APERR(Ec::InvalidCommand, "Job configuration is required");

    // Get the job config
    auto jobConfig = param.jobConfig.value().get();
    if (!jobConfig.isObject() || !jobConfig.isMember("taskId") ||
        !jobConfig["taskId"].isString())
        return APERR(Ec::InvalidCommand,
                     "taskId is required in your job configuration");

    // Get the task id
    auto taskId = jobConfig["taskId"].asString();

    // If there is no service config, use the null endpoint
    // Null is also a value, check for null explicitly
    if (!param.serviceConfig.has_value() ||
        param.serviceConfig.value().get().isNull())
        param.serviceConfig = defaultServiceConfig;

    // Get the type specified
    auto logicalType = IServiceEndpoint::getLogicalType(*param.serviceConfig);
    if (!logicalType) return logicalType.ccode();

    // Get the type specified
    auto physicalType = IServiceEndpoint::getPhysicalType(*param.serviceConfig);
    if (!physicalType) return physicalType.ccode();

    // Setup the pipe info
    IPipeType pipe{*logicalType, *logicalType, *physicalType};

    // Make the endpoint
    auto result = Factory::make<IServiceEndpoint>(_location, pipe);
    if (!result) return result.ccode();

    // Move it to a shared ptr
    ServiceEndpoint endpoint = _mv(*result);

    // Set the taskId
    endpoint->taskId = taskId;

    // Set the self referential weak ptr
    endpoint->endpoint = endpoint;

    // Save the endpoint mode
    endpoint->config.endpointMode = endpointMode;

    // Get our service definition if we have one, if we don't just
    // return the logical type
    auto service = IServices::getServiceDefinition(*logicalType);
    if (!service) return service.ccode();

    // Save the capabilities
    endpoint->capabilities = (*service)->capabilities;

    // Decode everything
    if (auto ccode = endpoint->setConfig(*param.jobConfig, *param.taskConfig,
                                         *param.serviceConfig))
        return ccode;

    // Save the request to the get pipe stack only - not begin it
    endpoint->m_stackOnly = param.stackOnly;

    // If we are in debug mode, bind the debugger
    if (param.debug) {
        // Bind it to this endpoint
        Debugger::registerDebugger(endpoint);
    }

    // If an open mode was specified, open it
    if (param.openMode != OPEN_MODE::NONE) {
        // Begin operations
        if (auto ccode = endpoint->beginEndpoint(param.openMode)) return ccode;

        // If we are a sync style driver, open the key value store
        if (endpoint->capabilities & Url::PROTOCOL_CAPS::SYNC) {
            // Key-Value storage
            if (auto ccode = endpoint->openKeyValueStorage()) return ccode;
        }
    }

    // And return the new endpoint
    return endpoint;
}

//-------------------------------------------------------------------------
/// @details
///		Static function to easily create an endpoint
///	@param[in]	param
///		The parameters for the endpoint
//-------------------------------------------------------------------------
ErrorOr<ServiceEndpoint> IServiceEndpoint::getSourceEndpoint(
    ENDPOINT_PARAM &&param) noexcept {
    return getEndpoint(ENDPOINT_MODE::SOURCE, param);
}

//-------------------------------------------------------------------------
/// @details
///		Static function to easily create an endpoint
///	@param[in]	param
///		The parameters for the endpoint
//-------------------------------------------------------------------------
ErrorOr<ServiceEndpoint> IServiceEndpoint::getTargetEndpoint(
    ENDPOINT_PARAM &&param) noexcept {
    return getEndpoint(ENDPOINT_MODE::TARGET, param);
}

//-------------------------------------------------------------------------
/// @details
///		Inserts a filter into the pipe. This is only valid during
///     beginGlobal
///	@param[in]	filter
///		Name of filter to instantiate
/// @param[in]  config
///     Configuration of the filter
///	@returns
///		Error
//-------------------------------------------------------------------------
Error IServiceEndpoint::insertFilter(const Text &filterName,
                                     const json::Value &filterConfig) {
    // Check to make sure this is a valid op
    if (pipeStackIndex < 0)
        throw APERR(Ec::InvalidCommand, "Only available during beginGlobal");

    // Create the filter object
    IPipeType filter;
    filter.id = filterName;
    filter.logicalType = filterName;
    filter.physicalType = filterName;
    filter.connConfig = filterConfig;

    // Ensure pipeStack is valid and insert the filter at the specified index
    if (pipeStackIndex < 0 || _cast<size_t>(pipeStackIndex) > pipeStack.size())
        throw APERR(Ec::OutOfRange, "Invalid pipe stack index");

    // Insert the filter at the current index
    pipeStack.insert(pipeStack.begin() + pipeStackIndex + 1, _mv(filter));

    // Increment the index after the insertion
    pipeStackIndex++;
    return {};
}

//-------------------------------------------------------------------------
/// @details
///		Open the endpoints key value store
///	@returns
///		Error
//-------------------------------------------------------------------------
Error IServiceEndpoint::openKeyValueStorage() noexcept {
    // get keystore parameter
    Text keyStoreUrlText;
    if (auto ccode =
            config.taskConfig.lookupAssign("keystore", keyStoreUrlText))
        return ccode;
    if (keyStoreUrlText.empty()) return {};

    // open keystore
    Url keyStoreUrl{keyStoreUrlText};
    Text subKey;
    if (auto ccode = getConfigSubKey(subKey)) return ccode;
    if (!subKey)
        return APERR(Ec::InvalidCommand,
                     "Configuration subkey is required to open key store");
    // Key format would consists from
    //  - the physical type of the endpoint
    //  - the logical type of the endpoint
    //  - the config subkey for the endpoint
    // For example, `python_ms-sharepoint_` where
    //  - `python` is the physical type of the endpoint
    //  - `ms-sharepoint` is the logical type of the endpoint
    //  - `` (empty string) is the config subkey for the endpoint (empty in case
    //  of SharePoint online)
    Text serviceKey =
        _ts(config.physicalType, "_", config.logicalType, "_", subKey);
    LOGT("Opening keystore with key", serviceKey);
    if (auto ccode = keystore::open(keyStoreUrl, serviceKey); ccode.hasCcode())
        return ccode.ccode();
    else
        m_keyStore = _mv(*ccode);

    return {};
}

}  // namespace engine::store